src/test/librados/io_cxx.cc

```diff
@@ -218,6 +218,41 @@ TEST_F(LibRadosIoPP, SparseReadOpPP) {
   }
 }
 
+TEST_F(LibRadosIoPP, SparseReadExtentArrayOpPP) {
+  char buf[128], zbuf[128];
+  memset(buf, 0xcc, sizeof(buf));
+  memset(zbuf, 0, sizeof(zbuf));
+  bufferlist bl;
+  int i, len = 1024, skip = 5;
+  bl.append(buf, sizeof(buf));
+  for (i = 0; i < len; i++) {
+    if (!(i % skip)) {
+      ASSERT_EQ(0, ioctx.write("foo4", bl, bl.length(), i * sizeof(buf)));
+    }
+  }
+
+  bufferlist expect_bl;
+  for (i = 0; i < len; i++) {
+    if (!(i % skip)) {
+      expect_bl.append(buf, sizeof(buf));
+    } else if (i <= len - skip) {
+      expect_bl.append(zbuf, sizeof(zbuf));
+    }
+  }
+
+  i = 0;
+  while (i++ < 128) {
```

Comment:
Yeah, correct.

Reply:
> Yeah, correct.

It sounds like you are saying that sending the same sparse-read op 128 times is intended.  What does that achieve/trigger on the OSD side?

Reply:
Sorry for confusing. This just will test `128` times for the sparse read. For each time the sprase read will return around `200` extents.

Reply:
> For each time the sprase read will return around `200` extents.

... and what is point of repeating that 128 times?  Are you expecting a different code path to be taken in the OSD on one of the iterations?

Reply:
Because reproducing this problem requires a certain probability. If this doesn't make sense I will remove it.

Reply:
> Because reproducing this problem requires a certain probability.

Could you please explain why?  What exactly needs to happen for the issue the reproduce and the test to fail?

I apologize for being dense, but I still don't understand what exactly is being fixed here.  I was hoping to derive that from the test but the test as presented is not reproducing anything for me either.

Reply:
When fail the `assert_eq_sparse(expect_bl, extents, read_bl);` check should fail.

The `extents array` will be incorrect and also the data decoded.

Reply:
> The `extents array` will be incorrect and also the data decoded.

And what needs to happen on the OSD side for that to occur?  Please be specific -- an explanation could look like:
1) the object layout is `[extentA, extentB, extentC]`, everything else is unallocated
2) client sends `sparse-read(off, len)` request against the object (if more than one request is needed, describe all)
3) OSD does X ... instead of Y ... because of Z
4) incorrect extents map and/or data is encoded by the OSD (describe what is actually encoded)
5) incorrect extents map and/or data is decoded by the client (ditto)

Again, I'm asking because I don't understand what exactly is being fixed here.  The PR essentially just gets rid of `std::move` in `src/osd/PrimaryLogPG.cc` but that doesn't change anything materially hence my confusion.

Reply:
I didn't dig it further to find out why and how the backstore could change the `m` map, I am not familiar with that.

But from the code and the comments I got that the `objects_readv_sync()` will possibly do that. 

If that happens so the `extents map` encoded for the sparse-read will be incorrect.

Reply:
> I didn't dig it further to find out why and how the backstore could change the `m` map, I am not familiar with that.

Ah!

> 
> But from the code and the comments I got that the `objects_readv_sync()` will possibly do that.

Yeah, the code in `PrimaryLogPG` is arranged to allow for that (but I don't think BlueStore ever does that).

> 
> If that happens so the `extents map` encoded for the sparse-read will be incorrect.

No, it won't be incorrect.  This is why I kept asking for a test or at least an in-depth explanation of what is being fixed.  Here is how `objects_readv_sync()` looks in main:
```
int ReplicatedBackend::objects_readv_sync(
  const hobject_t &hoid,
  map<uint64_t, uint64_t>&& m,
  uint32_t op_flags,
  bufferlist *bl)
{
  interval_set<uint64_t> im(std::move(m));
  auto r = store->readv(ch, ghobject_t(hoid), im, *bl, op_flags);
  if (r >= 0) {
    m = std::move(im).detach();
  }
  return r;
}
```
If `store->readv()` changes the interval set, those changes would make it back into the map with ` m = std::move(im).detach()`.  An rvalue reference (`&&`) is still a reference, so assignment still works.  The fact that in main `m` is "moved" into `objects_readv_sync()`:
```
    r = pgbackend->objects_readv_sync(soid, std::move(m), op.flags, &data_bl);
```
doesn't change anything -- remember that `std::move()` doesn't actually move but rather just casts its argument to an rvalue reference. 

> 
> And also locally I couldn't reproduce any issue with the current code by running the test for more than 1 days. And currently this test case is what I can figure out.

Then I think we should:
1) make it clear that this PR is just cleanup and nothing is actually fixed here
2) unlink this PR from https://tracker.ceph.com/issues/62081 (which is marked for backporting to all releases)
3) drop the test since it doesn't reproduce anything even in your environment

Reply:
I think you are right. 

Locally I did some test about this, something like:

`std::map<int, int> &&mapB = std::move(mapA);`

The `mapB` will also just be a reference to `mapA`. And it's totally different with:

`std::map<int, int> mapB = std::move(mapA);`

I think a test for `multiple extents` is necessary for `sparse-read`, but I will raise a new PR to do that.

For this PR I will just try to clean the code.

Reply:
> I think a test for `multiple extents` is necessary for `sparse-read`, but I will raise a new PR to do that.

Sounds good, please CC me!

> 
> For this PR I will just try to clean the code.

Need to drop `Fixes: https://tracker.ceph.com/issues/63318` from the commit message.

Reply:
> Need to drop `Fixes: https://tracker.ceph.com/issues/63318` from the commit message.

Ah, please disregard.  I see that it's a different tracker, specifically for "use-after-move".